src/data_source/storage/sql.rs

```diff
@@ -408,6 +426,135 @@ impl SqlStorage {
+                        Utc::now().timestamp() - (cfg.target_retention().as_secs()) as i64,
+                    )
+                    .await?;
+
+                if let Some(target_height) = target_height {
+                    while height < target_height {
+                        height = min(height + batch_size, target_height);
+                        self.query_opt("DELETE FROM header WHERE height < $1", &[&(height as i64)])
+                            .await?;
+                        tracing::info!("Pruned data up to height {}", height);
+                    }
+                    self.pruned_height = Some(target_height);
+                }
+
+                // If threshold is set, prune data exceeding minimum retention in batches
+                // This parameter is needed for SQL storage as there is no direct way to get free space.
+                if let Some(threshold) = cfg.pruning_threshold() {
+                    let mut usage = self.get_disk_usage().await?;
+
+                    // Prune data exceeding minimum retention in batches starting from minimum height
+                    // until usage is below threshold
+                    if usage > threshold {
+                        tracing::warn!(
+                            "Disk usage {} exceeds pruning threshold {:?}",
+                            usage,
+                            cfg.pruning_threshold()
+                        );
+                        let minimum_retention_height = self
+                            .get_height_by_timestamp(
+                                Utc::now().timestamp() - (cfg.minimum_retention().as_secs()) as i64,
+                            )
+                            .await?;
+
+                        if let Some(min_retention_height) = minimum_retention_height {
+                            while (usage as f64 / threshold as f64) > (max_usage as f64 / 10000.0)
+                                && height < min_retention_height
+                            {
+                                height = min(height + batch_size, min_retention_height);
+                                self.query_opt(
+                                    "DELETE FROM header WHERE height < $1",
+                                    &[&(height as i64)],
+                                )
+                                .await?;
+                                usage = self.get_disk_usage().await?;
+                                tracing::info!("Pruned data up to height {}", height);
+                            }
+                            self.pruned_height = Some(height);
+                        }
+                    }
+                }
+                // Vacuum the database to reclaim space.
+                // Note: VACUUM FULL is not used as it requires an exclusive lock on the tables, which can
+                // cause downtime for the query service.
+                self.query_opt_static("VACUUM;").await?;
```

Comment:
Since `VACUUM` doesn't return anything, you may have to use the low-level `self.client.batch_execute("VACUUM")` here